src/openssl/mod.rs

```diff
@@ -310,20 +325,358 @@ impl ECVRF {
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * multiplicative inverse of x in F
+    fn inv0(&mut self, x: &BigNum) -> Result<BigNum, Error> {
+        let two = BigNum::from_u32(2)?;
+
+        let mut qm2 = BigNum::new()?;
+        qm2.mod_sub(&self.p, &two, &self.p, &mut self.bn_ctx)?;
+
+        let mut result = BigNum::new()?;
+        result.mod_exp(x, &qm2, &self.p, &mut self.bn_ctx)?;
+        Ok(result)
+    }
+
+    /// Function to get the sign of a field element
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-4)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * -1 or 1
+    fn sgn0(&mut self, x: &BigNum) -> Result<i32, Error> {
+        // This function is not yet generalized for (m > 1)
+        let mut sign = 0i32;
+        let zero = BigNum::from_u32(0)?;
+        let one = BigNum::from_u32(1)?;
+        let two = BigNum::from_u32(2)?;
+
+        let mut pm1 = BigNum::new()?;
+        pm1.mod_sub(&self.p, &one, &self.p, &mut self.bn_ctx)?;
+        let mut inv2 = BigNum::new()?;
+        inv2.mod_inverse(&two, &self.p, &mut self.bn_ctx)?;
+        let mut pm1d2 = BigNum::new()?;
+        pm1d2.mod_mul(&pm1, &inv2, &self.p, &mut self.bn_ctx)?;
+
+        let criterion = x.ucmp(&pm1d2) == Ordering::Greater;
+        let mut sign_0 = ECVRF::icmov(1, -1, criterion);
+        sign_0 = ECVRF::icmov(sign_0, 0, x.ucmp(&zero) == Ordering::Equal);
+        sign = ECVRF::icmov(sign, sign_0, sign == 0);
+
+        Ok(ECVRF::icmov(sign, 1, sign == 0))
+    }
+
+    /// Function to get the value of squared y on the curve
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * the value of x^3 + a * x + b
+    fn function_g(&mut self, x: &BigNum) -> Result<BigNum, Error> {
+        let mut xsquare = BigNum::new()?;
+        xsquare.mod_sqr(x, &self.p, &mut self.bn_ctx)?;
+        let mut xqubic = BigNum::new()?;
+        xqubic.mod_mul(&xsquare, x, &self.p, &mut self.bn_ctx)?;
+
+        let mut ax = BigNum::new()?;
+        ax.mod_mul(x, &self.a, &self.p, &mut self.bn_ctx)?;
+
+        let mut xcubic_plus_ax = BigNum::new()?;
+        xcubic_plus_ax.mod_add(&xqubic, &ax, &self.p, &mut self.bn_ctx)?;
+        let mut ysquare = BigNum::new()?;
+        ysquare.mod_add(&xcubic_plus_ax, &self.b, &self.p, &mut self.bn_ctx)?;
+
+        Ok(ysquare)
+    }
+
+    /// Function to convert a `Hash(PK|DATA)` to a point in the curve, in constant time.
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-6.9.1)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `public_key` - An `EcPoint` referencing the public key.
+    /// * `alpha` - A slice containing the input data.
```

Comment:
There's no description about the argument `z`. I think `z` could be included in `CipherSuite`, not in the arguments here, because we know `self.cipher_suite` and the curve.

Reply:
Ok. It sounds great. I think I can introduce `fn z_vlaue` to `impl CipherSuite` similar to the `fn suite_string`

Reply:
I introduced the function to get z_value from `CipherSuite`s